### PR TITLE
[QueueRunner] Support "thenables" returned by tests

### PR DESCRIPTION
**Pull request by aleclarson**
This commit allows `done` or `done.fail` to be called by a returned promise/thenable.

This reduces the following boilerplate:
- `promise.then(done)`
- `promise.catch(done.fail)`

Includes **tests** for resolved promises and rejected promises.
Should I include one for pending promises, too?

### Example

```js
it(() => {
  return fetch('http://www.google.com')
    .then(res => expect(res).toBe(...));
});
```

- If the returned thenable is resolved, the `done` function is called.
- If the returned thenable is rejected, the `done.fail` function is called.

### Failure tests

To test failure cases, you can `catch` the error and call `expect(error.message)`.

```js
it(() => {
  return fetch('bad url')
    .catch(error => expect(error.message).toBe(...));
});
```

### CODE REVIEW

**Comment by aleclarson** [comment]
Am I not allowed to use `setImmediate` in the tests?

**Comment by aleclarson** [comment]
Fixed an edge case where the `promise` variable was being carried over to the next iteration, resulting in an error being thrown `TypeError: promise.then is not a function`.

Let me know if there are other issues. 👍

**Comment by BenoitZugmeyer** [comment]
Disclaimer: I'm not related to Jasmine in any way, just an enthusiastic user.

This feature is quite popular and has been discussed and PR'd many times (PR #202, #681, PR #696, #923, PR #1029), and has always been rejected with the following workarounds:

* You can just pass `done` and `done.fail` as callbacks to `.then`
* If you are into coroutines, you can use [jasmine-co](https://github.com/gradecam/jasmine-co)
* You can use [jasmine-promises](https://github.com/matthewjh/jasmine-promises)

In my opinion, it would be a great addition to Jasmine:
* Most other popular test runners (mocha, jest, ava, ...) are supporting asynchronous tests via promises out of the box. It seems like a new consensus, like the `done()` callback a few years ago. It would help migrations from one test runner to another, and build tools/helpers compatible with multiple test runners.
* Coroutines were alright a few years ago, but now that `async/await` functions are standardized and supported by Node (last stable version) and some browsers out of the box, there is not much reason to use them anymore. With support for tests returning promises, jasmine would support async functions without any extra code.
* The jasmine-promises doesn't support nodejs (at the moment) and seems a bit abandonned to be honnest (nodejs support is on the master branch, but no release has been done for six months).
* Adding promises support is straightforward and wouldn't add much code, and no extra dependency.

As a side note, @aleclarson : your PR breaks the CI build because of lints reported by jshint (https://travis-ci.org/jasmine/jasmine/jobs/202125138#L380)

**Comment by aleclarson** [comment]
@BenoitZugmeyer Thanks for the heads up! All fixed.

I'll squash the commits once a maintainer speaks up about any nits.

File: spec/core/QueueRunnerSpec.js [hunk 1]
```diff
@@ -234,6 +234,52 @@ describe("QueueRunner", function() {
       queueRunner.execute();
       expect(doneReturn).toBe(null);
     });
+
+    it("supports promises (and thenables) returned by queued functions", function(done) {
+      var queueableFn1 = { fn: function() {
+          return Promise.resolve();
```
Comment (slackersoft) [inline spec/core/QueueRunnerSpec.js]: Jasmine's tests need to run in IE 8+ and PhantomJS (as well as Edge, Chrome, Firefox, and Node.js) so we can't use bare `Promise` directly in specs for browsers that don't support them.

File: spec/core/QueueRunnerSpec.js [hunk 2]
```diff
@@ -234,6 +234,52 @@ describe("QueueRunner", function() {
...
+          queueableFns: [queueableFn1, queueableFn2],
+          onComplete: completeCallback
+        });
+
+      queueRunner.execute();
+
+      expect(completeCallback).not.toHaveBeenCalled();
+      expect(queueableFn2.fn).not.toHaveBeenCalled();
+
+      setImmediate(function() {
```
Comment (slackersoft) [inline spec/core/QueueRunnerSpec.js]: Similarly, `setImmediate` isn't available in all browsers that Jasmine supports and as such can't be used in Jasmine's tests.

File: spec/core/QueueRunnerSpec.js [hunk 3]
```diff
@@ -234,6 +234,52 @@ describe("QueueRunner", function() {
...
+        expect(completeCallback).toHaveBeenCalled();
+        expect(queueableFn2.fn).toHaveBeenCalled();
+        done();
+      });
+    });
+
+    it("calls 'done.fail' when a promise is rejected", function(done) {
+      var error, queueableFn1 = { fn: function() {
+          return Promise.reject(error = new Error('foo'));
```
Comment (slackersoft) [inline spec/core/QueueRunnerSpec.js]: Native `Promise` as above

File: spec/core/QueueRunnerSpec.js [hunk 4]
```diff
@@ -234,6 +234,52 @@ describe("QueueRunner", function() {
...
+          queueableFns: [queueableFn1, queueableFn2],
+          fail: failFn
+        });
+
+      queueRunner.execute();
+
+      expect(failFn).not.toHaveBeenCalled();
+      expect(queueableFn2.fn).not.toHaveBeenCalled();
+
+      setImmediate(function() {
```
Comment (slackersoft) [inline spec/core/QueueRunnerSpec.js]: `setImmediate` as above
